stl/src/filesystem.cpp

```diff
@@ -503,36 +499,124 @@ _Success_(return == __std_win_error::_Success) __std_win_error
     //     DWORD Flags;
     // } FILE_DISPOSITION_INFO_EX, *PFILE_DISPOSITION_INFO_EX;
 
-    struct _File_disposition_info_ex {
-        DWORD _Flags;
-    };
     _File_disposition_info_ex _Info_ex{0x3};
 
     // FileDispositionInfoEx isn't documented in MSDN at the time of this writing, but is present
     // in minwinbase.h as of at least 10.0.16299.0
     constexpr auto _FileDispositionInfoExClass = static_cast<FILE_INFO_BY_HANDLE_CLASS>(21);
-    if (SetFileInformationByHandle(_Handle._Get(), _FileDispositionInfoExClass, &_Info_ex, sizeof(_Info_ex))) {
-        return {true, __std_win_error::_Success};
+    if (SetFileInformationByHandle(
+            reinterpret_cast<HANDLE>(_Handle), _FileDispositionInfoExClass, &_Info_ex, sizeof(_Info_ex))) {
+        return __std_win_error::_Success;
     }
 
-    _Last_error = __std_win_error{GetLastError()};
+    const auto _Last_error = __std_win_error{GetLastError()};
     switch (_Last_error) {
     case __std_win_error::_Invalid_parameter: // Older Windows versions
     case __std_win_error::_Invalid_function: // Windows 10 1607
     case __std_win_error::_Not_supported: // POSIX delete not supported by the file system
         break; // try non-POSIX delete below
+    case __std_win_error::_Access_denied: // This might be due to the read-only bit, try to clear it and try again
     default:
-        return {false, _Last_error};
+        return _Last_error;
     }
 
     FILE_DISPOSITION_INFO _Info{/* .Delete= */ TRUE};
-    if (SetFileInformationByHandle(_Handle._Get(), FileDispositionInfo, &_Info, sizeof(_Info))) {
+    if (SetFileInformationByHandle(reinterpret_cast<HANDLE>(_Handle), FileDispositionInfo, &_Info, sizeof(_Info))) {
+        return __std_win_error::_Success;
+    }
+
+    return __std_win_error{GetLastError()};
+}
+
+[[nodiscard]] __std_fs_remove_result __stdcall __std_fs_remove(_In_z_ const wchar_t* const _Target) noexcept {
+    // remove _Target without caring whether _Target is a file or directory
+    __std_win_error _Last_error;
+    bool _Able_to_change_attributes = false;
+
+    constexpr auto _Flags = __std_fs_file_flags::_Backup_semantics | __std_fs_file_flags::_Open_reparse_point;
+    _STD _Fs_file _Handle(_Target,
+        __std_access_rights::_Delete | __std_access_rights::_File_read_attributes
+            | __std_access_rights::_File_write_attributes,
+        _Flags, &_Last_error);
+    if (_Last_error == __std_win_error::_Success) {
+        _Able_to_change_attributes = true;
+    } else if (_Last_error == __std_win_error::_Access_denied) {
+        // change the underlying HANDLE of _Handle
+        // this might not be a good way of doing it (comment to be removed post review)
+        _Last_error = __std_fs_open_handle(&_Handle._Raw, _Target, __std_access_rights::_Delete, _Flags);

```

Comment:
I *think* this is OK, because when `_Fs_file`'s constructor calls `__std_fs_open_handle` (which calls `__vcp_CreateFile`) and that fails, [the handle is](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew#return-value) `INVALID_HANDLE_VALUE`, so the destructor will do nothing. So, there's no concern about leaking a handle by overwriting `_Handle._Raw`.

I think this deserves a comment so retaining the first line `// change the underlying HANDLE of _Handle` would be good, although the second line should of course be removed.